src/lavinmq/mqtt/client.cr

```diff
@@ -179,15 +190,17 @@ module LavinMQ
 
       private def publish_will
         if will = @will
-          @broker.publish MQTT::Publish.new(
+          @broker.publish(@user, MQTT::Publish.new(
```

Comment:
Need to check write permissions for the vhost + exchange here as well and you can remove the rescue for AccessRefused